nixos/modules/services/networking/wireguard-networkd.nix

```
@@ -0,0 +1,181 @@
+      MTUBytes = interface.mtu;
+    };
+    wireguardConfig = removeNulls {
+      PrivateKeyFile = interface.privateKeyFile;
+      ListenPort = interface.listenPort;
+      FirewallMark = interface.fwMark;
+      RouteTable = if interface.allowedIPsAsRoutes then interface.table else null;
+      RouteMetric = interface.metric;
+    };
+    wireguardPeers = map generateWireguardPeer interface.peers;
+  };
+
+  generateWireguardPeer = peer: {
+    wireguardPeerConfig = removeNulls {
+      PublicKey = peer.publicKey;
+      PresharedKeyFile = peer.presharedKeyFile;
+      AllowedIPs = peer.allowedIPs;
+      Endpoint = peer.endpoint;
+      PersistentKeepalive = peer.persistentKeepalive;
+    };
+  };
+
+  generateNetwork = name: interface: {
+    matchConfig.Name = name;
+    address = interface.ips;
+  };
+
+  cfg = config.networking.wireguard;
+
+  refreshEnabledInterfaces = filterAttrs (name: interface: interface.dynamicEndpointRefreshSeconds != 0) cfg.interfaces;
+
+  generateRefreshTimer = name: interface: nameValuePair "wireguard-dynamic-refresh-${name}" {
+    partOf = [ "wireguard-dynamic-refresh-${name}.service" ];
+    wantedBy = [ "timers.target" ];
+    description = "Wireguard dynamic endpoint refresh (${name}) timer";
+    timerConfig.OnBootSec = interface.dynamicEndpointRefreshSeconds;
+    timerConfig.OnUnitInactiveSec = interface.dynamicEndpointRefreshSeconds;
+  };
+
+  generateRefreshService = name: interface: nameValuePair "wireguard-dynamic-refresh-${name}" {
+    description = "Wireguard dynamic endpoint refresh (${name})";
+    after = [ "network-online.target" ];
+    wants = [ "network-online.target" ];
+    path = with pkgs; [
+      iproute2
+      systemd
+    ];
+    # networkd doesn't provide a mechanism for refreshing endpoints.
+    # See: https://github.com/systemd/systemd/issues/9911
+    # This hack does the job but takes down the whole interface to do it.
+    script = ''
+      ip link delete ${name}
+      networkctl reload
+    '';
+  };
+
+in
+{
+  meta.maintainers = [ lib.maintainers.majiir ];
+
+  options.networking.wireguard = {
+    useNetworkd = mkOption {
+      default = config.networking.useNetworkd;
+      defaultText = literalExpression "config.networking.useNetworkd";
+      type = types.bool;
+      description = mdDoc ''
+        Whether to use networkd as the network configuration backend for
+        Wireguard instead of the legacy script-based system.
+
+        ::: {.warning}
+        Some options have slightly different behavior with the networkd and
+        script-based backends. Check the documentation for each Wireguard
+        option you use before enabling this option.
+        :::
+      '';
+    };
+  };
+
+  config = mkIf (cfg.enable && cfg.useNetworkd) {
+
+    # TODO: Some of these options may be possible to support in networkd.
+    #
+    # privateKey and presharedKey are trivial to support, but we deliberately
+    # don't to discourage putting secrets in the /nix store.
```

Initially, I read that as a double negation as I skipped the `to`.
Hopefully, the comma should prevent this for other readers

```suggestion
    # don't, to discourage putting secrets in the /nix store.
```

---

I don't think that would be grammatically correct, but I rephrased it to address the confusion:

> privateKey and presharedKey are trivial to support, but we deliberately don't **in order to** discourage putting secrets in the /nix store.